dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/preheat/TrackerPreheat.java

```diff
@@ -631,18 +632,17 @@ public Optional<TrackedEntityComment> getNote( String uid )
         return Optional.ofNullable( notes.get( uid ) );
     }
 
-    public Relationship getRelationship( org.hisp.dhis.tracker.domain.Relationship relationship )
+    public Relationship getRelationship( String relationshipUid )
     {
-        RelationshipType relationshipType = get( RelationshipType.class, relationship.getRelationshipType() );
+        return relationships.get( relationshipUid );
+    }
 
-        if ( relationship.getUid() != null && relationships.containsKey( relationship.getUid() ) )
-        {
-            return relationships.get( relationship.getUid() );
-        }
+    public Relationship getRelationshipByKey( org.hisp.dhis.tracker.domain.Relationship relationship )
```

Comment:
could this just be `getRelationshipByKey(Rel...)` ?
What is the meaning of `Key` here and in the other APIs you added?

Reply:
Now it should be clearer.
I separated completely the concept of preheated relationships and duplicated relationships.
The first ones are defined only by the UID, the second ones are defined by the key, that is a combination of the relationship type, from, to and the bidirectional flag